### DISCUSSION TITLE
mediatek/filogic预编译镜像中带nmbm字样的应该如何安装？

### DISCUSSION THREAD

**Discussion by Mosney** [immortalwrt/immortalwrt#1084] [Q&A] [answered]
https://downloads.immortalwrt.org/releases/23.05.0-rc4/targets/mediatek/filogic/
h3c nx30 pro和xiaomi wr30u都有，文件命名不太一样，initramfs-kernel、squashfs-factory、squashfs-sysupgrade扩展名都是直接用了bin，感觉用法上应该跟标准的itb fip不同，搜了半天也没搜到文档说明，openwrt主线那里也只有[标准版本](https://github.com/openwrt/openwrt/pull/12820)没有nmbm，就来这里问一问，我的设备是nx30pro

**Comment by KardRi** [comment]
我推測，網上最為廣傳的 237及酱紫表的教程，他們給的Uboot (checksum 5c12開頭) 是有NMBM的支持。很大的可能他們給的immortalwrt也是有nmbm的支持(否則你得刷BL2)

所以，刷入18.04 K2那個映像後，估計可以直接進luci sysupgrade刷入支援NMBM的image，或回到uboot刷

我沒試過，徵求勇者

註: WR30u nmbm的commit comment

> Because this device enable NMBM by default, most users use custom
> U-Boot with NMBM-Enabled in Chinese forums.
> 
> This layout is the same as the ubootmod layout but enabling NMBM.

**Reply by Mosney** [reply]
BL2这个分区是做什么的呢？我一直以为是bootloader2的缩写，但没搞懂它起什么作用，主线op和immortal都有一个额外preloader文件，按照 https://github.com/openwrt/openwrt/pull/12820 里a7s所说是要刷写BL2分区的，但是下面有人问BL2无法直接刷入时a7s说不刷BL2也行。酱紫表应该就是用的237的uboot，二者教程都没有提到需要刷写BL2

**Reply by KardRi** [reply]
我刷成功了，用酱紫表/237他們提供的U-boot

用他們提供的nx30pro.bin的immortal魔改版不能強制sysupgrade，我記得應該有那個選項的才對。

最後我是用u-boot刷 nmbm帶factory.bin的 X-wrt 
x-wrt-23.10-b202310140123-mediatek-filogic-h3c_magic-nx30-pro-nmbm-squashfs-factory.bin

channel 36 160Mhz 2.4Gbps正常

這個刷完大概還有36MB可以用，比起原本nx30pro.bin只有給16MB可以用多1倍

https://i.imgur.com/XCzRWnY.png

我個人猜測，要刷BL2是因為openwrt的固件沒有支援NMBM，所以要額外刷支持legacy版的Bootloader

又有想過，還是要刷BL2是因為他們的是大分區的image? 不過感覺不太可能，反正我是不願意去動BL2

**Reply by Mosney** [reply]
要想可用空间大的话最好还是从spi flash的分区下手，就是所谓的“大分区uboot版本”，默认都是跟原厂分区布局保持一致ubi只有64MB，最后可用空间就有限。
不过Immortal也是可以做到36MB可用空间的，237是因为预装了太多东西导致的可用空间相对小，而主线OpenWrt是多了几个ubi分区导致的可用空间同样相对小，多的ubi我也不清楚是做什么用的。另外我之前稀里糊涂的从op主线sysupgrqde刷到immortal可用空间就变大了，`ubinfo -a`看了下比op主线少了几个ubi分区

![IMG_20231014_221429_168](https://github.com/immortalwrt/immortalwrt/assets/22477230/4fe04412-e5b7-48de-b5d8-a2687e568ddb)

**Reply by KardRi** [reply]
我不知道差在哪裡，不熟，但這幾個NMBM相容的image，看mtd是都沒有變化的，都是9個
我有看到網上刷大分區的mtd5 ubi後面都不見了，變成config

```
root@X-WRT:~# df -h
Filesystem                Size      Used Available Use% Mounted on
/dev/root                13.3M     13.3M         0 100% /rom
devtmpfs                117.5M         0    117.5M   0% /rom/dev
tmpfs                   117.7M      1.1M    116.6M   1% /tmp
/dev/ubi0_2              37.2M    216.0K     35.0M   1% /overlay
overlayfs:/overlay       37.2M    216.0K     35.0M   1% /
tmpfs                     1.0M         0      1.0M   0% /mnt
tmpfs                   512.0K         0    512.0K   0% /dev
root@X-WRT:~#
```

<img width="513" alt="image" src="https://github.com/immortalwrt/immortalwrt/assets/51556723/9965221f-1a4b-4a2a-af4c-04fe9980615a">

**Reply by Mosney** [reply]
大分区版本就是把用不到的mtd6-9分区也利用起来。原厂是把/挂载到了tmpfs，用pdt_data储存配置信息，至于后面30多MB的plugin分区显然是为uu加速器等插件预留的，这些分区在openwrt中都是用不上的
```
root@NX30Pro:~# df -h
Filesystem                Size      Used Available Use% Mounted on
/dev/root                13.3M     13.3M         0 100% /rom
tmpfs                   113.2M      2.4M    110.8M   2% /tmp
tmpfs                   113.2M      1.3M    111.9M   1% /tmp/root
overlayfs:/tmp/root     113.2M      1.3M    111.9M   1% /
/dev/mtdblock6            6.0M      1.4M      4.6M  23% /mnt/config
/dev/mtdblock7            6.0M      1.4M      4.6M  23% /mnt/backup
tmpfs                   512.0K      4.0K    508.0K   1% /dev

```
```
dev:    size     erasesize   size       name
mtd0: 08000000   00020000   128.00 MB   "spi0.0"
mtd1: 00100000   00020000   1.00 MB     "BL2"
mtd2: 00080000   00020000   512.00 KB   "u-boot-env"
mtd3: 00200000   00020000   2.00 MB     "Factory"
mtd4: 00200000   00020000   2.00 MB     "FIP"
mtd5: 04000000   00020000   64.00 MB    "ubi"
mtd6: 00600000   00020000   6.00 MB     "pdt_data"
mtd7: 00600000   00020000   6.00 MB     "pdt_data_1"
mtd8: 00100000   00020000   1.00 MB     "exp"
mtd9: 02580000   00020000   38.25 MB    "plugin"

```
至于ubi分区看起来immortal和xwrt的ubi分区差不多
```
root@NX75:~# ubinfo -a
UBI version:                    1
Count of UBI devices:           1
UBI control device major/minor: 10:127
Present UBI devices:            ubi0
ubi0
Volumes count:                           3
Logical eraseblock size:                 126976 bytes, 124.0 KiB
Total amount of logical eraseblocks:     512 (65011712 bytes, 62.0 MiB)
Amount of available logical eraseblocks: 0 (0 bytes)
Maximum count of volumes                 128
Count of bad physical eraseblocks:       0
Count of reserved physical eraseblocks:  20
Current maximum erase counter value:     6
Minimum input/output unit size:          2048 bytes
Character device major/minor:            249:0
Present volumes:                         0, 1, 2
Volume ID:   0 (on ubi0)
Type:        dynamic
Alignment:   1
Size:        28 LEBs (3555328 bytes, 3.3 MiB)
State:       OK
Name:        kernel
Character device major/minor: 249:1
-----------------------------------
Volume ID:   1 (on ubi0)
Type:        dynamic
Alignment:   1
Size:        117 LEBs (14856192 bytes, 14.1 MiB)
State:       OK
Name:        fit
Character device major/minor: 249:2
-----------------------------------
Volume ID:   2 (on ubi0)
Type:        dynamic
Alignment:   1
Size:        341 LEBs (43298816 bytes, 41.2 MiB)
State:       OK
Name:        rootfs_data
Character device major/minor: 249:3
```

**Reply by KardRi** [reply]
話說，
哪裡可以下這行指令
`nmbm nmbm0 state`

起初，我很害怕如果我這台nand前幾個有壞塊，刷第三方U-boot直接一鍵變磚，尤其237沒說他的uboot有沒有nmbm的支持

看了一下，你我的壞塊都是0
`Count of bad physical eraseblocks:       0`
，但看了[主線](https://github.com/openwrt/openwrt/commit/dc2d4d73939c3d86a8e9d968c5c3462f92771bc6#diff-0a263858df577cc26996da59d6407fd35d3e8744454266ca76109ac7c6c29618) 的感覺，就是在賭前幾block不能有壞塊，尤其BL2有鎖的版本，若替換掉且無NMBM支持，offset可能會錯誤(?)

**Reply by Mosney** [reply]
不清楚，我还没用上nmbm，确实也有一台NX30Pro因为疑似坏快被我售后了，症状是OpenWrt下重复进recovery模式/刷回原厂后重启掉配置

**Reply by KardRi** [reply]
顯然immortalwrt NMBM也是可以uboot刷的，就是這個溫度，真的很感人
immortalwrt-23.05.0-rc4-mediatek-filogic-h3c_magic-nx30-pro-nmbm-squashfs-factory.bin

對了，我是剛到手的機器 8月份製造，原廠固件V100R002L02

[kernel log](https://pastebin.com/QTEe4L1V)

<img width="490" alt="image" src="https://github.com/immortalwrt/immortalwrt/assets/51556723/9b73ba9d-c12c-4e10-b38c-806e0bc1cd99">

mtd0原本是SPI整體，immortal似乎不見了，不影響就是

```
root@ImmortalWrt:~# cat /proc/mtd
dev:    size   erasesize  name
mtd0: 00100000 00020000 "BL2"
mtd1: 00080000 00020000 "u-boot-env"
mtd2: 00200000 00020000 "Factory"
mtd3: 00200000 00020000 "FIP"
mtd4: 04000000 00020000 "ubi"
mtd5: 00600000 00020000 "pdt_data"
mtd6: 00600000 00020000 "pdt_data_1"
mtd7: 00100000 00020000 "exp"
mtd8: 02580000 00020000 "plugin"
root@ImmortalWrt:~#
```

```
root@ImmortalWrt:~# df -h
Filesystem                Size      Used Available Use% Mounted on
/dev/root                 7.5M      7.5M         0 100% /rom
tmpfs                   115.1M    108.0K    115.0M   0% /tmp
/dev/ubi0_2              41.6M    168.0K     39.2M   0% /overlay
overlayfs:/overlay       41.6M    168.0K     39.2M   0% /
tmpfs                   512.0K         0    512.0K   0% /dev
```
```
root@ImmortalWrt:~# ubinfo -a
UBI version:                    1
Count of UBI devices:           1
UBI control device major/minor: 10:127
Present UBI devices:            ubi0

ubi0
Volumes count:                           3
Logical eraseblock size:                 126976 bytes, 124.0 KiB
Total amount of logical eraseblocks:     512 (65011712 bytes, 62.0 MiB)
Amount of available logical eraseblocks: 0 (0 bytes)
Maximum count of volumes                 128
Count of bad physical eraseblocks:       0
Count of reserved physical eraseblocks:  17
Current maximum erase counter value:     2
Minimum input/output unit size:          2048 bytes
Character device major/minor:            249:0
Present volumes:                         0, 1, 2

Volume ID:   0 (on ubi0)
Type:        dynamic
Alignment:   1
Size:        41 LEBs (5206016 bytes, 4.9 MiB)
State:       OK
Name:        kernel
Character device major/minor: 249:1
-----------------------------------
Volume ID:   1 (on ubi0)
Type:        dynamic
Alignment:   1
Size:        61 LEBs (7745536 bytes, 7.3 MiB)
State:       OK
Name:        rootfs
Character device major/minor: 249:2
-----------------------------------
Volume ID:   2 (on ubi0)
Type:        dynamic
Alignment:   1
Size:        387 LEBs (49139712 bytes, 46.8 MiB)
State:       OK
Name:        rootfs_data
Character device major/minor: 249:3
```

**Reply by Mosney** [reply]
~~用uboot刷完squashfs-factory是不是recovery模式？还需要sysupgrade刷一下squashfs-sysupgrade？~~

**Reply by KardRi** [reply]
不需要，Uboot刷完nmbm-squashfs-factory.bin直接就能用，不必再sysupgrade一次

只是如果你先前有刷過BL2，那勸你不要刷NMBM layout

其實我覺得如果H3C原廠Uboot有tftp recovery，那應該不用刷Uboot，感覺

**Reply by Mosney** [reply]
为什么刷过BL2就不要用NMBM？

**Comment by Mosney** [comment]
看了一下文件命名和大小，我觉得应该是
`nmbm-initramfs-kernel.bin`对应`initramfs-recovery.itb`

但是`nmbm-squashfs-factory.bin`和`nmbm-squashfs-sysupgrade.bin`有什么区别暂且不明，factory的意思可能是原厂分区布局？但是看起来后者看起来才更像是对应`squashfs-sysupgrade.itb`的. 

还得请contributor @1715173329 来答疑解惑一下了：
1. 上述两个不同尾缀的文件区别是什么？什么情况下该用哪个？
2. 使用nmbm版本是否需要相应的uboot版本配合，抑或是uboot通用？

**Comment by Mosney** [comment]
**记录一下尝试过程，目前稳定在1使用：**
1. 用[237的uboot](https://sssddddff.lanzoup.com/igjBl10tm6va)，[来源](https://www.right.com.cn/forum/thread-8291820-1-1.html)，uboot webui上传并写入`nmbm-squashfs-factory.bin`，此时可用空间41MiB，随后luci下sysupgrade刷写`nmbm-squashfs-sysupgrade.bin`
    <details>
      <summary>ubinfo如下</summary>
    
    ```
    
    root@ImmortalWrt:~# ubinfo -a
    UBI version:                    1
    Count of UBI devices:           1
    UBI control device major/minor: 10:127
    Present UBI devices:            ubi0
    
    ubi0
    Volumes count:                           3
    Logical eraseblock size:                 126976 bytes, 124.0 KiB
    Total amount of logical eraseblocks:     512 (65011712 bytes, 62.0 MiB)
    Amount of available logical eraseblocks: 0 (0 bytes)
    Maximum count of volumes                 128
    Count of bad physical eraseblocks:       0
    Count of reserved physical eraseblocks:  19
    Current maximum erase counter value:     2
    Minimum input/output unit size:          2048 bytes
    Character device major/minor:            249:0
    Present volumes:                         0, 1, 2
    
    Volume ID:   0 (on ubi0)
    Type:        dynamic
    Alignment:   1
    Size:        41 LEBs (5206016 bytes, 4.9 MiB)
    State:       OK
    Name:        kernel
    Character device major/minor: 249:1
    -----------------------------------
    Volume ID:   1 (on ubi0)
    Type:        dynamic
    Alignment:   1
    Size:        61 LEBs (7745536 bytes, 7.3 MiB)
    State:       OK
    Name:        rootfs
    Character device major/minor: 249:2
    -----------------------------------
    Volume ID:   2 (on ubi0)
    Type:        dynamic
    Alignment:   1
    Size:        385 LEBs (48885760 bytes, 46.6 MiB)
    State:       OK
    Name:        rootfs_data
    Character device major/minor: 249:3
    ```
    
    </details>


2. 使用上述uboot也可以直接刷写`nmbm-squashfs-sysupgrade.bin`，但可用空间会比前述少很多只有16MiB左右，ubi会有6个Volumes，并且还有10MiB+未分配
    <details>
      <summary>ubinfo如下</summary>
      
    ```
    root@NX30Pro:~# ubinfo -a
    UBI version:                    1
    Count of UBI devices:           1
    UBI control device major/minor: 10:62
    Present UBI devices:            ubi0
    
    ubi0
    Volumes count:                           6
    Logical eraseblock size:                 126976 bytes, 124.0 KiB
    Total amount of logical eraseblocks:     512 (65011712 bytes, 62.0 MiB)
    Amount of available logical eraseblocks: 85 (10792960 bytes, 10.2 MiB)
    Maximum count of volumes                 128
    Count of bad physical eraseblocks:       0
    Count of reserved physical eraseblocks:  19
    Current maximum erase counter value:     2
    Minimum input/output unit size:          2048 bytes
    Character device major/minor:            250:0
    Present volumes:                         0, 1, 2, 3, 4, 5
    
    Volume ID:   0 (on ubi0)
    Type:        dynamic
    Alignment:   1
    Size:        25 LEBs (3174400 bytes, 3.0 MiB)
    State:       OK
    Name:        kernel
    Character device major/minor: 250:1
    -----------------------------------
    Volume ID:   1 (on ubi0)
    Type:        dynamic
    Alignment:   1
    Size:        141 LEBs (17903616 bytes, 17.0 MiB)
    State:       OK
    Name:        rootfs
    Character device major/minor: 250:2
    -----------------------------------
    Volume ID:   2 (on ubi0)
    Type:        dynamic
    Alignment:   1
    Size:        25 LEBs (3174400 bytes, 3.0 MiB)
    State:       OK
    Name:        kernel2
    Character device major/minor: 250:3
    -----------------------------------
    Volume ID:   3 (on ubi0)
    Type:        dynamic
    Alignment:   1
    Size:        141 LEBs (17903616 bytes, 17.0 MiB)
    State:       OK
    Name:        rootfs2
    Character device major/minor: 250:4
    -----------------------------------
    Volume ID:   4 (on ubi0)
    Type:        dynamic
    Alignment:   1
    Size:        67 LEBs (8507392 bytes, 8.1 MiB)
    State:       OK
    Name:        rootfs_data
    Character device major/minor: 250:5
    -----------------------------------
    Volume ID:   5 (on ubi0)
    Type:        dynamic
    Alignment:   1
    Size:        5 LEBs (634880 bytes, 620.0 KiB)
    State:       OK
    Name:        u-boot-env
    Character device major/minor: 250:6
    ```
    
    </details>


3. 另外还尝试在luci下使用sysupgrade上传`nmbm-squashfs-factory.bin`，但会有如下错误提示，可以推测出squashfs-factory不是给sysupgrade用的
   > 镜像检查失败了：
   > upgrade: Image metadata not present
   > upgrade: Use sysupgrade -F to override this check when downgrading or flashing to vendor firmware Image check failed. 

   不过既然可以uboot可以直接刷`nmbm-squashfs-factory.bin` 那么`nmbm-initramfs-kernel`好像就没什么作用了，反之如果uboot理应刷入`nmbm-initramfs-kernel`那`nmbm-squashfs-factory.bin`又是做什么用的呢？

**Reply by Mosney** [reply, marked as answer]
根据[这里](https://firmware-selector.immortalwrt.org/?version=23.05.0-rc4&target=mediatek%2Ffilogic&id=h3c_magic-nx30-pro-nmbm)的描述：

KERNEL	
内置最简文件系统的 Linux 内核，适用于首次安装或故障恢复。
FACTORY	
首次刷机时，使用 Factory 映像以刷入 ImmortalWrt。通常您可以在原厂固件的 Web 界面中完成此操作。
SYSUPGRADE	
使用 Sysupgrade 映像以更新现有运行 ImmortalWrt 的设备。该映像可以在 LuCI 界面或终端中使用。



ref:
https://openwrt.org/faq/what_is_the_difference_between_the_different_image_formats
https://openwrt.org/docs/techref/image.format#initramfs
https://forum.openwrt.org/t/what-is-initramfs-kernel/48561/3
https://www.ptt.cc/bbs/Broad_Band/M.1697255775.A.5BF.html

**Reply by Mosney** [reply]
实践了一下，237的uboot只能直接刷写factory，主线的uboot只能通过tftp先引导initramfs再sysupgrade写入内置flash

**Comment by hcwhan** [comment]
@Mosney 请问你在刷nmbm之前有刷非nmbm版本吗 我现在是已经刷了非nmbm版本
BL2和FIP分区已经刷了 ImmortalWrt的preloader.bin和bl31-uboot.fip 现在准备把FIP分区刷回237的uboot然后刷 nmbm版 不知道BL2分区需不需要刷回原厂

**Reply by Mosney** [reply]
我有尝试过从非NMBM直接sysupgrade到NMBM，记得是没有问题，不过为了保险起见我是恢复了之前备份的BL2分区的。后来FIP也换成了237的uboot，重新通过uboot刷写的NMBM版。

**Reply by Masked-cat** [reply]
实测Mosney老哥的方法可行，如果BL2和FIP分区已经刷了 ImmortalWrt的preloader.bin和bl31-uboot.fip，可以直接强制刷immortalwrt-xxx-mediatek-filogic-h3c_magic-nx30-pro-nmbm-squashfs-sysupgrade.bin固件，可以正常升级到NMBM版本。
